src/helpers.coffee

```diff
@@ -114,6 +114,17 @@ buildLocationData = (first, last) ->
       last.range[1]
     ]
 
+# Build a list of all comments attached to tokens.
+exports.extractAllCommentTokens = (tokens) ->
```

Comment:
I didn't literally reuse `buildTokenDataDictionary()` since it was more straightforward to just directly generate a sorted list of comments (rather than nest them/duplicate them under top-level token-associated keys)

But I guess if it was desirable to not do the additional loop over the tokens array, it might be possible to reuse the single `buildTokenDataDictionary()` data structure and extract the comments from there (depending on where we actually pass the comments to the AST root, per the previous comment)

Reply:
I'm not sure the best way to accurately time the performance, but when I just ran `time bin/cake build:except-parser` a few times with/without this loop included, there wasn't a noticeable difference between the two (there was significantly more variability between individual test runs for a given version than between the two versions, and there were individual passes of the version that included the loop that were faster than individual passes of the version that didn't include the loop)

And this additional loop is only being performed when they're building AST - I agree that it'd be nice to try and optimize the speed of the AST build for the reasons you mentioned. But I guess I'd picture that as a separate compiler-wide optimization pass once we have the AST functionality landed. I'd be curious to dig in and get a sense of how much time is typically spent in each of the compiler phases (eg rewriter passes)